src/Http/Routing/src/ValidationEndpointFilterFactory.cs

```diff
@@ -0,0 +1,100 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.ComponentModel.DataAnnotations;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Http.Validation;
+
+internal static class ValidationEndpointFilterFactory
+{
+    private const string ValidationContextJustification = "The DisplayName property is always statically initialized in the ValidationContext through this codepath.";
+
+    public static EndpointFilterDelegate Create(EndpointFilterFactoryContext context, EndpointFilterDelegate next)
+    {
+        var parameters = context.MethodInfo.GetParameters();
+        var options = context.ApplicationServices.GetService<IOptions<ValidationOptions>>()?.Value;
+        if (options is null)
+        {
+            return next;
+        }
+
+        var parameterCount = parameters.Length;
+        var validatableParameters = new IValidatableInfo[parameterCount];
+        var parameterDisplayNames = new string[parameterCount];
+        var hasValidatableParameters = false;
+
+        for (var i = 0; i < parameterCount; i++)
+        {
+            if (options.TryGetValidatableParameterInfo(parameters[i], out var validatableParameter))
+            {
+                validatableParameters[i] = validatableParameter;
+                parameterDisplayNames[i] = GetDisplayName(parameters[i]);
+                hasValidatableParameters = true;
+            }
+        }
+
+        if (!hasValidatableParameters)
+        {
+            return next;
+        }
+
+        var validatableContext = new ValidateContext { ValidationOptions = options };
+        return async (context) =>
+        {
+            validatableContext.ValidationErrors?.Clear();
```

Comment:
Since the context is created once per endpoint at startup (or first request), that means it's being used in parallel by all requests to the endpoint it's initialized for?